File: python/day11/day11.py
#input_file = "test-input11.txt"
#input_file = "python/day11/test-input11.txt" #for debugging from the AOC-Java-DJC workspace
input_file = "input11.txt"

"""
Cellular automata
"""
width = 0
height = 0
with open(input_file) as file:
    
    raw = file.read().split("\n")
    width = len(raw[0])
    height = len(raw)
    spots = []
    for row in raw:
        spots.append(list(row)) # manage strings as char lists


def part1(spots):
    """ 
    Find how many seats are occupied once the arrangement is stable using these
    rules :
    1> is empty ('L') and all adjacent are empty -> becomes occupied
    2> is occupied ('#') and 4+ adjacent are occupied -> becomes empty
    3> otherwise, no change

    Floor ('.') counts as a constant empty

    test-input11 -> should stabilize at 37 occupied seats after 6 loops
    """
    first = True
    loops = 0
    # need to declare and fill with each element completely 
    # independantly from other lists - including other rows
    # because assignment links to the same object in memory and 
    # changing one changes all the others that are linked
    next_spots = []
    for i in range(height):
        next_spots.append([]) # new empty list
        for j in range(width): next_spots[i].append('.') # item assignment, not list assignment
    

    while next_spots != spots: # until stable
        if not first:
            for i in range(height):
                for j in range(width): spots[i][j] = next_spots[i][j]
        first = False
        loops += 1

        for i in range(height):
            for j in range(width):
                # skip floor spots
                if spots[i][j] == '.': continue
                
                # get empty seats
                local_occupied = 0
                if i > 0 and j > 0 and spots[i-1][j-1] == '#': 
                    local_occupied += 1
                if i > 0 and spots[i-1][j] == '#':
                    local_occupied += 1
                if i > 0 and j < width - 1 and spots[i-1][j+1] == '#': 
                    local_occupied += 1
                if j < width - 1 and spots[i][j+1] == '#': 
                    local_occupied += 1
                if i < height - 1 and j < width - 1 and spots[i+1][j+1] == '#': 
                    local_occupied += 1
                if i < height - 1 and spots[i+1][j] == '#': 
                    local_occupied += 1
                if i < height - 1 and j > 0 and spots[i+1][j-1] == '#': 
                    local_occupied += 1
                if j > 0 and spots[i][j-1] == '#': 
                    local_occupied += 1
                
                # apply rules to seat
                if spots[i][j] == 'L' and local_occupied == 0: next_spots[i][j] = '#'
                elif spots[i][j] == '#' and local_occupied >= 4: next_spots[i][j] = 'L'
                else: next_spots[i][j] = spots[i][j]

    # count occupied seats
    occupied = 0
    for  r in next_spots: 
        for c in r:
            if c == '#': occupied += 1
    
    print(f"{loops} loops, {occupied} occupied seats")


    


def part2():
    """
    """
    pass


if __name__ == "__main__":
    part1(spots)
    part2()

File: days/Day11/Day11.java
import java.util.Arrays;

public class Day11 extends Day {
    String[] input;
    int width;
    int height;
    char[][] spots;
    
    Day11() {
        //cache input
        inputFile = "days/Day11/test-input11.txt";
        //inputFile = "days/Day11/input11.txt";
        input = getInput().split("\n");
        height = input.length;
        width = input[0].length();
        spots = new char[height][width];
        for (int i=0; i < input.length; i++) spots[i] = input[i].toCharArray();
    }

    /**
    Find how many seats are occupied once the arrangement is stable using these rules )
    1> is empty && all adjacent are empty -> becomes occupied
    2> is occupied && 4+ are occupied -> becomes empty
    3> otherwise, no change

    Presume floor counts as a constant empty

    test-input11 -> stabilizes at 37 seats after 6 loops

    ***** ACTUALLY HARDER TO MANAGE THAN THE PYTHON LISTS!!! ****
    I thought char arrays would be more well-behaved, but they seem to have
    the same issues :
     > assignment issues (values versus objects)
     > dependancies of values within a table (changing one item modifies multiple linked items)
     */
    public void part1() 
    {
        boolean first = true;
        int loops = 0;
        // linked objects, cannot change one without affecting the other
        char[][] nextSpots = spots;
        /* // seemingly independant items -> algo doesn't converge
        char[][] nextSpots = new char[height][width];
        for (int i = 0; i < height; i++){
            for (int j = 0; j < width; j++) nextSpots[i][j] = '.';
        }
        */
        /* // another seemingly independant declaration -> algo also hangs
        char[][] nextSpots = new char[height][width];
        for (int i = 0; i < height; i++){
            nextSpots[i] = Arrays.copyOf(spots[i], width);
        }
        */

        while (nextSpots != spots || first)
        {
            if (!first)
            {
                for (int i = 0; i < height; i++)
                {
                    for (int j = 0; j < width; j++) spots[i][j] = nextSpots[i][j]; // assigning items, not arrays (pointers to objects)
                }
            }
            first = false;
            loops++;

            for (int i = 0; i < height; i++)
            {
                for (int j = 0; j < width; j++)
                {
                    // skip floor spots
                    if (spots[i][j] == '.') continue;

                    // scan occupied seats
                    int localOccupied = 0;
                    if (i > 0 && j > 0 && spots[i-1][j-1] == '#') localOccupied++;
                    if (i > 0 && spots[i-1][j] == '#') localOccupied++;
                    if (i > 0 && j < width - 1 && spots[i-1][j+1] == '#') localOccupied++;
                    if (j < width - 1 && spots[i][j+1] == '#') localOccupied++;
                    if (i < height - 1 && j < width - 1 && spots[i+1][j+1] == '#') localOccupied++;
                    if (i < height - 1 && spots[i+1][j] == '#') localOccupied++;
                    if (i < height - 1 && j > 0 && spots[i+1][j-1] == '#') localOccupied++;
                    if (j > 0 && spots[i][j-1] == '#') localOccupied++;

                    // apply rules to the seat
                    if (spots[i][j] == 'L' && localOccupied == 0) nextSpots[i][j] = '#';
                    else if (localOccupied >= 4) nextSpots[i][j] = 'L';
                    else nextSpots[i][j] = spots[i][j];
                }
            }
        }
        
        int occupied = 0;
        for (char[] r : nextSpots)
        {
            for (char c : r) occupied += c == '#' ? 1 : 0;
        }
        System.out.printf("%d loops, %d occupied seats\n", loops, occupied);
    }


    /**
     * 
     */
    public void part2() {
        
        System.out.printf("\n");
    }
}